File: sdk/kotlin/src/main/kotlin/com/seocheon/sdk/modules/NodeModule.kt
package com.seocheon.sdk.modules

import com.seocheon.sdk.infrastructure.ChainClient
import com.seocheon.sdk.infrastructure.signing.SigningService
import com.seocheon.sdk.internal.tx.MsgConfirmDelegation
import com.seocheon.sdk.internal.tx.Pipeline
import com.seocheon.sdk.internal.tx.PipelineConfig
import com.seocheon.sdk.internal.tx.TxRequest
import com.seocheon.sdk.types.DelegationStatusResponse
import com.seocheon.sdk.types.NodeInfoResponse
import com.seocheon.sdk.types.NodeSearchResponse
import com.seocheon.sdk.types.NodeSummary
import com.seocheon.sdk.types.TxResultResponse
import kotlinx.serialization.json.*

/**
 * Node module for querying node information and searching.
 */
class NodeModule(
    private val client: ChainClient,
    private val signer: SigningService,
    private val pipelineConfig: PipelineConfig? = null,
) {

    /**
     * Retrieves detailed information for a node.
     */
    suspend fun getInfo(nodeId: String? = null): NodeInfoResponse {
        val id = nodeId ?: resolveOwnNodeId()
        val resp = client.queryRest("/seocheon/node/v1/nodes/$id")
        val node = resp.jsonObject["node"]?.jsonObject ?: resp.jsonObject

        return NodeInfoResponse(
            nodeId = node["id"]?.jsonPrimitive?.content ?: id,
            operator = node["operator"]?.jsonPrimitive?.content ?: "",
            agentAddress = node["agent_address"]?.jsonPrimitive?.content ?: "",
            status = node["status"]?.jsonPrimitive?.content ?: "UNSPECIFIED",
            description = node["description"]?.jsonPrimitive?.content ?: "",
            website = node["website"]?.jsonPrimitive?.content ?: "",
            tags = node["tags"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList(),
            commissionRate = node["commission_rate"]?.jsonPrimitive?.content ?: "0",
            agentShare = node["agent_share"]?.jsonPrimitive?.content ?: "0",
            totalDelegation = node["total_delegation"]?.jsonPrimitive?.content ?: "0",
            selfDelegation = node["self_delegation"]?.jsonPrimitive?.content ?: "0",
            validatorAddress = node["validator_address"]?.jsonPrimitive?.content ?: "",
            registeredAt = node["registered_at"]?.jsonPrimitive?.longOrNull ?: 0L,
        )
    }

    /**
     * Searches for nodes with optional filters.
     */
    suspend fun search(
        tag: String? = null,
        status: String? = null,
        limit: Int = 20,
        orderBy: String = "delegation",
    ): NodeSearchResponse {
        val params = mutableListOf<String>()
        tag?.let { params.add("tag=$it") }
        status?.let { params.add("status=$it") }
        params.add("limit=$limit")
        params.add("order_by=$orderBy")

        val query = if (params.isNotEmpty()) "?${params.joinToString("&")}" else ""
        val resp = client.queryRest("/seocheon/node/v1/nodes$query")

        val nodes = resp.jsonObject["nodes"]?.jsonArray?.map { item ->
            val obj = item.jsonObject
            NodeSummary(
                nodeId = obj["id"]?.jsonPrimitive?.content ?: "",
                status = obj["status"]?.jsonPrimitive?.content ?: "UNSPECIFIED",
                tags = obj["tags"]?.jsonArray?.map { it.jsonPrimitive.content } ?: emptyList(),
                totalDelegation = obj["total_delegation"]?.jsonPrimitive?.content ?: "0",
                description = obj["description"]?.jsonPrimitive?.content ?: "",
            )
        } ?: emptyList()

        val totalCount = resp.jsonObject["total_count"]?.jsonPrimitive?.longOrNull ?: nodes.size.toLong()
        return NodeSearchResponse(nodes = nodes, totalCount = totalCount)
    }

    /**
     * Queries the delegation confirmation status.
     */
    suspend fun getDelegationStatus(delegatorAddress: String, validatorAddress: String): DelegationStatusResponse {
        val resp = client.queryRest("/seocheon/node/v1/delegation-confirmation/$delegatorAddress/$validatorAddress")
        val obj = resp.jsonObject
        return DelegationStatusResponse(
            expiryEpoch = obj["expiry_epoch"]?.jsonPrimitive?.longOrNull ?: 0L,
            currentEpoch = obj["current_epoch"]?.jsonPrimitive?.longOrNull ?: 0L,
            inRenewalWindow = obj["in_renewal_window"]?.jsonPrimitive?.booleanOrNull ?: false,
            renewalWindowStart = obj["renewal_window_start"]?.jsonPrimitive?.longOrNull ?: 0L,
        )
    }

    /**
     * Confirms delegation to a validator.
     */
    suspend fun confirmDelegation(validatorAddress: String): TxResultResponse {
        val cfg = pipelineConfig ?: throw com.seocheon.sdk.errors.SdkError.InvalidConfig("PipelineConfig required for TX operations")
        val msg = MsgConfirmDelegation(
            delegatorAddress = signer.getAddress(),
            validatorAddress = validatorAddress,
        )
        val result = Pipeline.executeTx(client, signerAdapter(), cfg, TxRequest(message = msg))
        return TxResultResponse(
            txHash = result.txHash,
            height = result.height,
            code = result.code,
            gasUsed = result.gasUsed,
            gasWanted = result.gasWanted,
            rawLog = result.rawLog,
            events = emptyList(),
        )
    }

    private fun signerAdapter(): com.seocheon.sdk.internal.tx.Signer = object : com.seocheon.sdk.internal.tx.Signer {
        override suspend fun sign(data: ByteArray): ByteArray = signer.sign(data)
        override fun getAddress(): String = signer.getAddress()
        override fun getPubKey(): ByteArray = signer.getPubKey()
    }

    private suspend fun resolveOwnNodeId(): String {
        val address = signer.getAddress()
        val resp = client.queryRest("/seocheon/node/v1/nodes/by-agent/$address")
        return resp.jsonObject["node"]?.jsonObject?.get("id")?.jsonPrimitive?.content
            ?: throw com.seocheon.sdk.errors.SdkError.NodeNotFound()
    }
}

File: sdk/csharp/src/Modules/NodeModule.cs
using Seocheon.Sdk.Errors;
using Seocheon.Sdk.Infrastructure;
using Seocheon.Sdk.Infrastructure.Signing;
using Seocheon.Sdk.Internal.Tx;
using Seocheon.Sdk.Types;

namespace Seocheon.Sdk.Modules;

/// <summary>
/// Node module: query node information and search.
/// </summary>
public sealed class NodeModule
{
    private readonly IChainClient _client;
    private readonly ISigningService _signer;
    private readonly PipelineConfig _pipelineConfig;

    internal NodeModule(IChainClient client, ISigningService signer, PipelineConfig? pipelineConfig = null)
    {
        _client = client;
        _signer = signer;
        _pipelineConfig = pipelineConfig ?? new PipelineConfig { ChainId = "seocheon-1" };
    }

    /// <summary>
    /// Returns detailed information for a specific node.
    /// </summary>
    /// <param name="nodeId">Node ID. Empty resolves from signer's agent address.</param>
    /// <param name="ct">Cancellation token.</param>
    public async Task<NodeInfoResponse> GetInfo(string nodeId = "", CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(nodeId))
            nodeId = await ResolveOwnNodeId(ct);

        var result = await _client.QueryRest($"/seocheon/node/v1/nodes/{nodeId}", ct);
        var node = result.GetProperty("node");

        var tags = new List<string>();
        if (node.TryGetProperty("tags", out var tagsEl))
        {
            foreach (var t in tagsEl.EnumerateArray())
                tags.Add(t.GetString() ?? "");
        }

        return new NodeInfoResponse(
            NodeId: node.GetProperty("id").GetString() ?? "",
            Operator: node.TryGetProperty("operator", out var op) ? op.GetString() ?? "" : "",
            AgentAddress: node.TryGetProperty("agent_address", out var aa) ? aa.GetString() ?? "" : "",
            Status: node.TryGetProperty("status", out var st) ? st.GetString() ?? "REGISTERED" : "REGISTERED",
            Description: node.TryGetProperty("description", out var desc) ? desc.GetString() ?? "" : "",
            Website: node.TryGetProperty("website", out var ws) ? ws.GetString() ?? "" : "",
            Tags: tags,
            CommissionRate: node.TryGetProperty("commission_rate", out var cr) ? cr.GetString() ?? "0" : "0",
            AgentShare: node.TryGetProperty("agent_share", out var @as) ? @as.GetString() ?? "0" : "0",
            TotalDelegation: node.TryGetProperty("total_delegation", out var td) ? td.GetString() ?? "0" : "0",
            SelfDelegation: node.TryGetProperty("self_delegation", out var sd) ? sd.GetString() ?? "0" : "0",
            ValidatorAddress: node.TryGetProperty("validator_address", out var va) ? va.GetString() ?? "" : "",
            RegisteredAt: node.TryGetProperty("registered_at", out var ra)
                ? long.Parse(ra.GetString() ?? "0") : 0
        );
    }

    /// <summary>
    /// Searches for nodes with optional filters.
    /// </summary>
    /// <param name="tag">Filter by tag. Empty for all.</param>
    /// <param name="status">Filter by status. Empty for all.</param>
    /// <param name="limit">Maximum results (default 20).</param>
    /// <param name="orderBy">"delegation" (default) or "registered_at".</param>
    /// <param name="ct">Cancellation token.</param>
    public async Task<NodeSearchResponse> Search(
        string tag = "",
        string status = "",
        int limit = 20,
        string orderBy = "delegation",
        CancellationToken ct = default)
    {
        var query = $"/seocheon/node/v1/nodes?limit={limit}&order_by={orderBy}";
        if (!string.IsNullOrEmpty(tag))
            query += $"&tag={Uri.EscapeDataString(tag)}";
        if (!string.IsNullOrEmpty(status))
            query += $"&status={Uri.EscapeDataString(status)}";

        var result = await _client.QueryRest(query, ct);

        var nodes = new List<NodeSummary>();
        if (result.TryGetProperty("nodes", out var nodesEl))
        {
            foreach (var n in nodesEl.EnumerateArray())
            {
                var tags = new List<string>();
                if (n.TryGetProperty("tags", out var tagsEl))
                {
                    foreach (var t in tagsEl.EnumerateArray())
                        tags.Add(t.GetString() ?? "");
                }

                nodes.Add(new NodeSummary(
                    NodeId: n.GetProperty("id").GetString() ?? "",
                    Status: n.TryGetProperty("status", out var st) ? st.GetString() ?? "" : "",
                    Tags: tags,
                    TotalDelegation: n.TryGetProperty("total_delegation", out var td) ? td.GetString() ?? "0" : "0",
                    Description: n.TryGetProperty("description", out var desc) ? desc.GetString() ?? "" : ""
                ));
            }
        }

        ulong total = 0;
        if (result.TryGetProperty("total_count", out var tc))
            ulong.TryParse(tc.GetString(), out total);

        return new NodeSearchResponse(nodes, total);
    }

    /// <summary>
    /// Queries delegation confirmation status.
    /// </summary>
    public async Task<DelegationStatusResponse> GetDelegationStatus(
        string delegatorAddress, string validatorAddress, CancellationToken ct = default)
    {
        var result = await _client.QueryRest(
            $"/seocheon/node/v1/delegation-confirmation/{delegatorAddress}/{validatorAddress}", ct);

        return new DelegationStatusResponse(
            ExpiryEpoch: result.TryGetProperty("expiry_epoch", out var ee) ? long.Parse(ee.GetString() ?? "0") : 0,
            CurrentEpoch: result.TryGetProperty("current_epoch", out var ce) ? long.Parse(ce.GetString() ?? "0") : 0,
            InRenewalWindow: result.TryGetProperty("in_renewal_window", out var rw) && rw.GetBoolean(),
            RenewalWindowStart: result.TryGetProperty("renewal_window_start", out var rws) ? long.Parse(rws.GetString() ?? "0") : 0
        );
    }

    /// <summary>
    /// Confirms delegation for a validator.
    /// </summary>
    public async Task<TxResultResponse> ConfirmDelegation(string validatorAddress, CancellationToken ct = default)
    {
        var msgBytes = Messages.EncodeMsgConfirmDelegation(
            _signer.GetAddress(),
            validatorAddress
        );

        var result = await Pipeline.ExecuteTx(_client, _signer, _pipelineConfig, new TxRequest
        {
            Message = msgBytes,
            MessageTypeUrl = Messages.TypeMsgConfirmDelegation
        }, ct);

        return new TxResultResponse(
            TxHash: result.TxHash,
            Height: result.Height,
            Code: result.Code,
            GasUsed: result.GasUsed,
            GasWanted: result.GasWanted,
            RawLog: result.RawLog,
            Events: Array.Empty<TxEvent>()
        );
    }

    private async Task<string> ResolveOwnNodeId(CancellationToken ct)
    {
        var address = _signer.GetAddress();
        var result = await _client.QueryRest($"/seocheon/node/v1/nodes/by-agent/{address}", ct);
        return result.GetProperty("node").GetProperty("id").GetString()
               ?? throw SdkErrors.QueryFailed("Could not resolve node ID");
    }
}
